operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java

```diff
@@ -0,0 +1,125 @@
+package org.keycloak.operator;
+
+import io.fabric8.kubernetes.api.model.EnvVarBuilder;
+import io.quarkus.logging.Log;
+import io.quarkus.test.junit.QuarkusTest;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.keycloak.operator.utils.CRAssert;
+import org.keycloak.operator.v2alpha1.crds.Keycloak;
+import org.keycloak.operator.v2alpha1.crds.KeycloakStatusCondition;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.keycloak.operator.utils.K8sUtils.deployKeycloak;
+import static org.keycloak.operator.utils.K8sUtils.getDefaultKeycloakDeployment;
+import static org.keycloak.operator.utils.K8sUtils.waitForKeycloakToBeReady;
+
+@QuarkusTest
+public class KeycloakDeploymentE2EIT extends ClusterOperatorTest {
+    @Test
+    public void testBasicKeycloakDeploymentAndDeletion() {
+        try {
+            // CR
+            Log.info("Creating new Keycloak CR example");
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            var deploymentName = kc.getMetadata().getName();
+            deployKeycloak(k8sclient, kc, true);
+
+            // Check Operator has deployed Keycloak
+            Log.info("Checking Operator has deployed Keycloak deployment");
+            assertThat(k8sclient.apps().deployments().inNamespace(namespace).withName(deploymentName).get()).isNotNull();
+
+            // Check Keycloak has correct replicas
+            Log.info("Checking Keycloak pod has ready replicas == 1");
+            assertThat(k8sclient.apps().deployments().inNamespace(namespace).withName(deploymentName).get().getStatus().getReadyReplicas()).isEqualTo(1);
+
+            // Delete CR
+            Log.info("Deleting Keycloak CR and watching cleanup");
+            k8sclient.resources(Keycloak.class).delete(kc);
+            Awaitility.await()
+                    .untilAsserted(() -> assertThat(k8sclient.apps().deployments().inNamespace(namespace).withName(deploymentName).get()).isNull());
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    @Test
+    public void testCRFields() {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            var deploymentName = kc.getMetadata().getName();
+            deployKeycloak(k8sclient, kc, true);
+
+            kc.getSpec().setImage("quay.io/keycloak/non-existing-keycloak");
+            kc.getSpec().getServerConfiguration().put("KC_DB_PASSWORD", "Ay Caramba!");
+            deployKeycloak(k8sclient, kc, false);
+
+            Awaitility.await()
+                    .during(Duration.ofSeconds(15)) // check if the Deployment is stable
```

Comment:
Why 15 seconds ? the benefit of using Awaitility is not having blocking delays and exit as soon as the value is right. 
What if in 5 seconds the value is the expected ? it will still wait 10 seconds more to do the first assert.

Even you are not specifying a final time right ? so it will be here for the eternity ?

Reply:
This is to make sure the deployment is stable and no more modified back to the original values by the  reconciler loop. The timeout is testsuite wise 180 seconds.